### ISSUE TITLE
RGB converts to grayscale by taking the R channel, in contradiction to what the doc states

### ISSUE THREAD

**Issue by bhaller** [lvandeve/lodepng#139]
I’m puzzled by what seems to be a mismatch between what the documentation says versus what the code is doing, specifically with conversion of an RGB image into grayscale during decode.  The docs in lodepng.h says this:

It can convert from almost any color type to any other color type, except the
following conversions: RGB to grayscale is not supported, and converting to a
palette when the palette doesn't have a required color is not supported. This is
not supported on purpose: this is information loss which requires a color
reduction algorithm that is beyond the scope of a PNG encoder (yes, RGB to gray
is easy, but there are multiple ways if you want to give some channels more
weight).

So from the phrase “RGB to grayscale is not supported”, I expected trying to decode an RGB image as grayscale to give back an error.  But I’ve got a trivial 7x5 pixel RGB PNG I just made (I can send it to you if you want me to, but didn’t attach it here since I didn’t want to risk triggering a spam filter or something), and when I request for it to be decoded as a grayscale image, the decode succeeds, with the red channel of the RGB image being taken as the grayscale values.  The relevant code in my app is:

	std::vector<unsigned char> png_data;
	unsigned width, height, error;
	
	error = lodepng::load_file(png_data, file_path);
	if (error)
		...;
	
	error = lodepng::decode(pixels_, width, height, png_data, LodePNGColorType::LCT_GREY, 8);

This call succeeds even though the image is RGB.  In the debugger, I stepped into the code to this backtrace:

#0	rgba8ToPixel(unsigned char*, unsigned long, LodePNGColorMode const*, ColorTree*, unsigned char, unsigned char, unsigned char, unsigned char) at lodepng.cpp:3160
#1	lodepng_convert(unsigned char*, unsigned char const*, LodePNGColorMode const*, LodePNGColorMode const*, unsigned int, unsigned int) at lodepng.cpp:3574
#2	lodepng_decode(unsigned char**, unsigned int*, unsigned int*, LodePNGState*, unsigned char const*, unsigned long) at lodepng.cpp:4955
#3	lodepng_decode_memory(unsigned char**, unsigned int*, unsigned int*, unsigned char const*, unsigned long, LodePNGColorType, unsigned int) at lodepng.cpp:4969
#4	lodepng::decode(std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >&, unsigned int&, unsigned int&, unsigned char const*, unsigned long, LodePNGColorType, unsigned int) at lodepng.cpp:6312
#5	lodepng::decode(std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >&, unsigned int&, unsigned int&, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > const&, LodePNGColorType, unsigned int) at lodepng.cpp:6326

The code at this point is:

static unsigned rgba8ToPixel(unsigned char* out, size_t i,
                            const LodePNGColorMode* mode, ColorTree* tree /*for palette*/,
                            unsigned char r, unsigned char g, unsigned char b, unsigned char a) {
 if(mode->colortype == LCT_GREY) {
   unsigned char gray = r; /*((unsigned short)r + g + b) / 3u;*/
   if(mode->bitdepth == 8) out[i] = gray;
   else if(mode->bitdepth == 16) out[i * 2 + 0] = out[i * 2 + 1] = gray;
   else {
     /*take the most significant bits of gray*/
     gray = ((unsigned)gray >> (8u - mode->bitdepth)) & ((1u << mode->bitdepth) - 1u);
     addColorBits(out, i, mode->bitdepth, gray);
   }
 } else ...

When mode->colortype == LCT_GREY, this seems to be taking the red channel of the RGB data and using it for the gray value, in “unsigned char gray = r;”.  It seems like, from what the documentation says, that whole “if” block ought to just return an error.

I was hoping to get an error back, since I want to decode images as 8-bit grayscale *if* they’re grayscale, otherwise as 24-bit RGB.  My strategy is thus to try the decode first as grayscale, and if that fails, try again as RGB.  But since the decode as grayscale is succeeding even when the image is RGB, this strategy isn’t working.

So my question is: is this a bug in LodePNG, or am I misunderstanding the documentation / misusing the library?  If the latter, how would you recommend that I use the library correctly, in order to get an 8-bit grayscale decode if-and-only-if the image is grayscale (of either 8 or 16 bit depth)?

By the way, I wonder if perhaps the bug is at line 4943:

   /*TODO: check if this works according to the statement in the documentation: "The converter can convert
   from grayscale input color type, to 8-bit grayscale or grayscale with alpha"*/
   if(!(state->info_raw.colortype == LCT_RGB || state->info_raw.colortype == LCT_RGBA)
      && !(state->info_raw.bitdepth == 8)) {
     return 56; /*unsupported color mode conversion*/
   }

This seems to be where unsupported color conversions get caught, but it doesn’t check for the case of converting RGB to grayscale.  I also note that there’s an error code defined for the RGB-to-grayscale case:

   case 62: return "conversion from color to grayscale not supported";

but nowhere in the code is error 62 generated.  So this seems to just be an oversight…?

I'd be happy to submit a PR to fix this bug, if that would be welcome.

**Comment by lvandeve** [comment]
Hi,

The error checking for this was removed at some point for performance reasons (not doing one more check per pixel), taking the red channel automatically works correctly if the image is grayscale. You have a good point that something needs to be clarified here though and silently just taking the red channel is not always desirable.

For implementing what you'd like: It's better not to rely on the error codes for this, since they are only meant to generate a human-readable message with the function returning strings.

You can use the function lodepng_inspect to check whether the original PNG image is grayscale or RGB beforehand, this checks only the header so is cheap, and then decode to the one of the other format based on this (if the images use the grayscale color type and aren't storing grayscale data in an RGB png).

Another option is to disable color_convert, then you get the result in any possible format the PNG could have been (1-bit, palette, RGBA, ...) which requires more handling of multiple cases of course

**Comment by bhaller** [comment]
Hi @lvandeve.  Agreed that a per-pixel check would be bad, for performance reasons.  To fix the problem in my copy of the code, I added this code in `lodepng_decode()`:

	/* BCH 11 Oct. 2020: check for conversion from color to grayscale, and refuse to do it */
	/* the documentation implies that this should happen, but error 62 is not thrown anywhere in the code */
	if ((state->info_raw.colortype == LCT_GREY || state->info_raw.colortype == LCT_GREY_ALPHA)
		&& (state->info_png.color.colortype == LCT_RGB || state->info_png.color.colortype == LCT_RGBA)) {
		return 62; /*conversion from color to grayscale not supported*/
	}

That does just a single check and returns error 62 if converting from RGB to grayscale has been requested.  It seems to work quite nicely.  I do think that using the red channel as "gray" should not be the default behavior; it works fine, of course, but is rather surprising (and if it is the standard behavior, then it should be documented, of course).  Perhaps it should be provided as an option (in which case it would presumably be useful to be able to request the green or blue channels instead), but it seems to me the default behavior of `lodepng_decode()` should be to give an error, as the documentation currently suggests.

That said, yes, your suggested strategy of using `lodepng_inspect()` first seems like it would work fine too.  If you decide that that should be the standard strategy for my use case – rather than having `lodepng_decode()` return an error for the RGB->grayscale conversion by default – then perhaps it would make a nice example in the documentation?  :->

Anyway, that's all the input I have on this bug, I think; do whatever you think is best.  :->  Thanks for the reply!

**Comment by lvandeve** [comment]
Hey, the case where the color is encoded RGB, but all three channels are equal (so, grayscale encoded in a less optimal way, it can occur sometimes), should be allowed, so the proposed colortype check in lodepng_decode would be too strict.

But with lodepng_inspect you can do exactly this, that's exactly an intended use case of it

**Comment by bhaller** [comment]
It should be *allowed*, sure – I'm just not sure that it ought to be the *default*, as opposed to being an option selected with a flag in the LodePNGState.  But if you decide you like the current behavior, that is of course your call; then I guess this is a documentation bug.  :->

**Comment by lvandeve** [comment]
Yeah, I'll improve something about it to avoid the trap, keeping this bug open, to improve with better documentation or an efficient form of error checking or both
